src/uct/tcp/tcp_ep.c

```diff
@@ -991,22 +1192,110 @@ ucs_status_t uct_tcp_ep_am_zcopy(uct_ep_h uct_ep, uint8_t am_id, const void *hea
+    uct_tcp_ep_zcopy_tx_t *ctx;
+    size_t payload_length;
+    ucs_status_t status;
+
+    status = uct_tcp_ep_prepare_zcopy(iface, ep, am_id, header, header_length,
+                                      iov, iovcnt, "am_zcopy",
+                                      iface->config.rx_seg_size -
+                                      sizeof(uct_tcp_am_hdr_t),
+                                      &payload_length, &hdr);
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ctx         = ucs_derived_of(hdr, uct_tcp_ep_zcopy_tx_t);
+    hdr->length = payload_length + header_length;
 
     status = uct_tcp_ep_am_sendv(iface, ep, 0, hdr,
                                  iface->config.rx_seg_size,
                                  header, ctx->iov, ctx->iov_cnt);
-    if ((status == UCS_OK) || (status == UCS_ERR_NO_PROGRESS)) {
-        UCT_TL_EP_STAT_OP(&ep->super, AM, ZCOPY, hdr->length);
+    if (ucs_unlikely((status != UCS_OK) && (status != UCS_ERR_NO_PROGRESS))) {
+        goto out;
+    }
 
-        if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
-            uct_tcp_ep_set_outstanding_zcopy(iface, ep, ctx, header,
-                                             header_length, comp);
-            return UCS_INPROGRESS;
-        }
+    UCT_TL_EP_STAT_OP(&ep->super, AM, ZCOPY, hdr->length);
+
+    if (uct_tcp_ep_ctx_buf_need_progress(&ep->tx)) {
+        uct_tcp_ep_set_outstanding_zcopy(iface, ep, ctx, header,
+                                         header_length, comp);
+        return UCS_INPROGRESS;
     }
 
+    ucs_assert(status == UCS_OK);
+
+out:
+    uct_tcp_ep_ctx_reset(&ep->tx);
+    return status;
+}
+
+ucs_status_t uct_tcp_ep_put_zcopy(uct_ep_h uct_ep, const uct_iov_t *iov,
+                                  size_t iovcnt, uint64_t remote_addr,
+                                  uct_rkey_t rkey, uct_completion_t *comp)
+{
+    uct_tcp_ep_t *ep             = ucs_derived_of(uct_ep, uct_tcp_ep_t);
+    uct_tcp_iface_t *iface       = ucs_derived_of(uct_ep->iface, uct_tcp_iface_t);
+    uct_tcp_am_hdr_t *hdr        = NULL;
+    uct_tcp_ep_put_req_t put_req = { 0 };
+    uct_tcp_ep_zcopy_tx_t *ctx;
+    ucs_status_t status;
+
+    status = uct_tcp_ep_prepare_zcopy(iface, ep, 0, &put_req, sizeof(put_req),
+                                      iov, iovcnt, "put_zcopy",
+                                      UCT_TCP_EP_PUT_ZCOPY_MAX -
+                                      sizeof(uct_tcp_am_hdr_t),
+                                      /* Set a payload length directly to the
+                                       * TX length, since PUT Zcopy doesn't
+                                       * set the payload length to TCP AM hdr */
+                                      &ep->tx.length, &hdr);
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ctx            = ucs_derived_of(hdr, uct_tcp_ep_zcopy_tx_t);
+    hdr->am_id     = UCT_TCP_EP_PUT_REQ_AM_ID;
```

Comment:
can't we pass this to uct_tcp_ep_prepare_zcopy instead of 0?

Reply:
unfortunately, no
`uct_tcp_ep_am_prepare` calls `UCT_CHECK_AM_ID`

Reply:
we can move AM_ID check from `uct_tcp_ep_am_prepare ` to the caller of `uct_tcp_ep_am_prepare` code. is it OK?